layer/cache_sideload_layer.cc

```diff
@@ -0,0 +1,427 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <cstring>
+#include <optional>
+#include <string>
+#include <string_view>
+
+#include "absl/container/flat_hash_map.h"
+#include "input_buffer.h"
+#include "layer_data.h"
+#include "layer_utils.h"
+#include "logging.h"
+
+namespace performancelayers {
+namespace {
+class CacheSideloadLayerData : public LayerData {
+ public:
+  CacheSideloadLayerData(const char* pipeline_cache_path)
+      : LayerData(nullptr, ""),
+        implicit_pipeline_cache_path_(pipeline_cache_path) {
+    LogEventOnly("cache_sideload_layer_init");
+  }
+
+  VkPipelineCache GetImplicitDeviceCache(VkDevice) const;
+  VkPipelineCache CreateImplicitDeviceCache(
+      VkDevice device, const VkAllocationCallbacks* alloc_callbacks,
+      absl::Span<const uint8_t> initial_data);
+
+  std::optional<InputBuffer> ReadImplicitCacheFile();
+
+ private:
+  mutable absl::Mutex device_to_implicit_cache_handle_lock_;
+  absl::flat_hash_map<VkDevice, VkPipelineCache>
+      device_to_implicit_cache_handle_
+          ABSL_GUARDED_BY(device_to_implicit_cache_handle_lock_);
+
+  const char* implicit_pipeline_cache_path_ = nullptr;
+};
+
+VkPipelineCache CacheSideloadLayerData::GetImplicitDeviceCache(
+    VkDevice device) const {
+  absl::MutexLock lock(&device_to_implicit_cache_handle_lock_);
+  if (auto it = device_to_implicit_cache_handle_.find(device);
+      it != device_to_implicit_cache_handle_.end())
+    return it->second;
+
+  return nullptr;
+}
+
+VkPipelineCache CacheSideloadLayerData::CreateImplicitDeviceCache(
+    VkDevice device, const VkAllocationCallbacks* alloc_callbacks,
+    absl::Span<const uint8_t> initial_data) {
+  absl::MutexLock lock(&device_to_implicit_cache_handle_lock_);
+  assert(device_to_implicit_cache_handle_.count(device) == 0 &&
```

Comment:
If `absl::MutexLock` is an inter-process lock, `assert` failure of a process can put another process in the deadlock.
Is it ok?

Reply:
This is not an inter-process lock. It's local to the layer. Layer functions can be called concurrently, but the lock is not used by any other code. When the assertion fails, it will crash the whole application. This is the intended behavior.